Removing multiple completely yields runtime error

**What happened**:
Removing multiple links completely with a rewrite rule fails with runtime error. 

```
panic: runtime error: slice bounds out of range [:104] with capacity 64

goroutine 54 [running]:
github.com/gardener/docforge/pkg/markdown/parser.(*link).GetDestination(0xc0004180c0, 0xc0004101b7, 0x1, 0x9)
        /home/i024136/dev/docforge/pkg/markdown/parser/links.go:115 +0x8b
github.com/gardener/docforge/pkg/markdown.UpdateMarkdownLinks.func1(0xc1b180, 0xc0004180c0)
        /home/i024136/dev/docforge/pkg/markdown/links.go:79 +0xbd
github.com/gardener/docforge/pkg/markdown/parser.(*document).ListLinks(0xc000400060, 0xc00040c030)
        /home/i024136/dev/docforge/pkg/markdown/parser/document.go:15 +0x82
github.com/gardener/docforge/pkg/markdown.UpdateMarkdownLinks(0xc00040a000, 0xa9, 0x2a9, 0xc000418000, 0x0, 0xc00040a000, 0xa9, 0x2a9, 0x41)
        /home/i024136/dev/docforge/pkg/markdown/links.go:67 +0xc4
github.com/gardener/docforge/pkg/reactor.(*nodeContentProcessor).reconcileMDLinks(0xc0000bf380, 0xc15700, 0xc0000cb640, 0xc0001be3f0, 0xc00040a000, 0xa9, 0x2a9, 0xc0000a8150, 0x24, 0x0, ...)
        /home/i024136/dev/docforge/pkg/reactor/content_processor.go:104 +0xe9
github.com/gardener/docforge/pkg/reactor.(*nodeContentProcessor).ReconcileLinks(0xc0000bf380, 0xc15700, 0xc0000cb640, 0xc0001be3f0, 0xc0000a8150, 0x24, 0xc00040a000, 0xa9, 0x2a9, 0x0, ...)
        /home/i024136/dev/docforge/pkg/reactor/content_processor.go:87 +0x1cd
github.com/gardener/docforge/pkg/reactor.(*DocumentWorker).Work(0xc0001771f0, 0xc15700, 0xc0000cb640, 0xa23ce0, 0xc0000bc348, 0xc15940, 0xc0001731a0, 0x2)
        /home/i024136/dev/docforge/pkg/reactor/document_worker.go:142 +0x8ee
github.com/gardener/docforge/pkg/jobs.(*Job).startWorkers.func1(0xc00017d220, 0xc000304000, 0xc15700, 0xc0000cb640, 0x4, 0xc15940, 0xc0001731a0, 0xc000302000)
        /home/i024136/dev/docforge/pkg/jobs/jobs.go:248 +0x302
created by github.com/gardener/docforge/pkg/jobs.(*Job).startWorkers
        /home/i024136/dev/docforge/pkg/jobs/jobs.go:232 +0x115
```
**What you expected to happen**:
Links to be removed

**How to reproduce it (as minimally and precisely as possible)**:
- manifest:
```
structure:
- source: /home/user/test/links.md
links:
  rewrites:
    (pulls|pull|issues|issue|wiki): ~
```
- markdown ( /home/user/test/links.md) with multiple links that match rewrite rule that removes them:
```
[a](https://github.com/gardener/documentation/wiki/Architecture)
[b](https://github.com/gardener/gardener/issues/356)
[c](https://github.com/gardener/gardener/pull/2309)
```

**How to categorize this issue?**
/kind bug
/priority critical

--- a/pkg/markdown/links.go
+++ b/pkg/markdown/links.go
@@ -64,7 +64,7 @@ func UpdateMarkdownLinks(documentBlob []byte, callback UpdateMarkdownLink) ([]by
 	if callback == nil {
 		return nil, nil
 	}
-	document.ListLinks(func(l parser.Link) {
+	err := document.ListLinks(func(l parser.Link) (parser.Link, error) {
 		var (
 			destination, text, title []byte
 			err                      error
@@ -77,17 +77,20 @@ func UpdateMarkdownLinks(documentBlob []byte, callback UpdateMarkdownLink) ([]by
 		}
 		text = l.GetText()
 		if destination, text, title, err = callback(t, l.GetDestination(), text, l.GetTitle()); err != nil {
-			return
+			return nil, err
 		}
-		updateLink(l, destination, text, title)
+		return updateLink(l, destination, text, title), nil
 	})
-	return document.Bytes(), nil
+	if err != nil {
+		return nil, err
+	}
+	return document.Bytes(), err
 }
 
-func updateLink(link parser.Link, destination, text, title []byte) {
+func updateLink(link parser.Link, destination, text, title []byte) parser.Link {
 	if destination == nil {
 		link.Remove(text != nil && len(text) > 0)
-		return
+		return nil
 	}
 	if text != nil && !bytes.Equal(link.GetText(), text) {
 		link.SetText(text)
@@ -98,4 +101,5 @@ func updateLink(link parser.Link, destination, text, title []byte) {
 	if title != nil && !bytes.Equal(link.GetTitle(), title) {
 		link.SetTitle(title)
 	}
+	return link
 }

--- a/pkg/markdown/parser/document.go
+++ b/pkg/markdown/parser/document.go
@@ -9,12 +9,27 @@ type document struct {
 	links []Link
 }
 
-func (d *document) ListLinks(cb UpdateMarkdownLinkListed) {
+func (d *document) ListLinks(cb OnLinkListed) error {
 	if cb != nil && d.links != nil {
+		i := 0
 		for _, l := range d.links {
-			cb(l)
+			_l, err := cb(l)
+			if err != nil {
+				return err
+			}
+			if _l != nil {
+				// Copy over only links that were not removed
+				d.links[i] = _l
+				i++
+			}
 		}
+		// Erase truncated values
+		for j := i; j < len(d.links); j++ {
+			d.links[j] = nil
+		}
+		d.links = d.links[:i]
 	}
+	return nil
 }
 
 func (d *document) Bytes() []byte {

--- a/pkg/markdown/parser/links.go
+++ b/pkg/markdown/parser/links.go
@@ -142,9 +142,9 @@ func (l *link) Remove(leaveText bool) {
 	l.document.data = append(l.document.data, text...)
 	l.document.data = append(l.document.data, doc2...)
 	var (
-		needsOffset         bool
-		offset              int
-		removedElementIndex int
+		offset int
+		_l     *link
+		ok     bool
 	)
 	if len(text) > 0 {
 		offset = len(text) - (l.end - l.start)
@@ -152,20 +152,14 @@ func (l *link) Remove(leaveText bool) {
 		offset = l.start - l.end
 	}
 	for i := 0; i < len(l.document.links); i++ {
-		if l.document.links[i] == l {
-			removedElementIndex = i
-			needsOffset = true
+		if _l, ok = l.document.links[i].(*link); !ok {
 			continue
 		}
-		if needsOffset {
-			offsetLinkByteRanges(l.document.links[i].(*link), offset)
+		// links positioned after the removed one get offset
+		if l.end < _l.start {
+			offsetLinkByteRanges(_l, offset)
 		}
 	}
-	l.document.links = remove(l.document.links, removedElementIndex)
-}
-
-func remove(slice []Link, s int) []Link {
-	return append(slice[:s], slice[s+1:]...)
 }
 
 func (l *link) IsImage() bool {

--- a/pkg/markdown/parser/links_test.go
+++ b/pkg/markdown/parser/links_test.go
@@ -14,7 +14,7 @@ import (
 func TestListLinkRewrites(t *testing.T) {
 	testCases := []struct {
 		in     *document
-		listCb UpdateMarkdownLinkListed
+		listCb OnLinkListed
 		want   []byte
 	}{
 		{
@@ -47,7 +47,7 @@ func TestListLinkRewrites(t *testing.T) {
 					},
 				},
 			},
-			func(l Link) {
+			func(l Link) (Link, error) {
 				if bytes.Equal(l.GetText(), []byte("b")) {
 					l.SetDestination([]byte("e.com"))
 					l.SetText([]byte("d"))
@@ -55,6 +55,7 @@ func TestListLinkRewrites(t *testing.T) {
 				if bytes.Equal(l.GetText(), []byte("b1")) {
 					l.SetDestination([]byte("c2.com"))
 				}
+				return l, nil
 			},
 			[]byte("some intro [d](e.com) some text [b1](c2.com)"),
 		},
@@ -88,7 +89,7 @@ func TestListLinkRewrites(t *testing.T) {
 					},
 				},
 			},
-			func(l Link) {
+			func(l Link) (Link, error) {
 				if bytes.Equal(l.GetText(), []byte("b")) {
 					l.SetDestination([]byte("e.com"))
 					l.SetText([]byte("d"))
@@ -96,6 +97,7 @@ func TestListLinkRewrites(t *testing.T) {
 				if bytes.Equal(l.GetText(), []byte("b1")) {
 					l.SetDestination([]byte("c2.com"))
 				}
+				return l, nil
 			},
 			[]byte("some intro [d](e.com)\n`x`some text [b1](c2.com)"),
 		},
@@ -130,7 +132,7 @@ func TestListLinkRewrites(t *testing.T) {
 					},
 				},
 			},
-			func(l Link) {
+			func(l Link) (Link, error) {
 				if bytes.Equal(l.GetText(), []byte("b")) {
 					l.SetDestination([]byte("e.com"))
 					l.SetText([]byte("d"))
@@ -138,13 +140,14 @@ func TestListLinkRewrites(t *testing.T) {
 				if bytes.Equal(l.GetText(), []byte("b1")) {
 					l.SetDestination([]byte("c2.com"))
 				}
+				return l, nil
 			},
 			[]byte(`some intro [d](e.com)
    some text [b1](c2.com)`),
 		},
 		{
 			&document{
-				[]byte("some intro [b](c.com)\n`x`some text [b1](c1.com)"),
+				[]byte("some intro [b](c.com)\n`x`some text [b1](c1.com)\n[b2](c2.com)"),
 				[]Link{
 					&link{
 						start: 11,
@@ -170,18 +173,32 @@ func TestListLinkRewrites(t *testing.T) {
 							end:   46,
 						},
 					},
+					&link{
+						start: 48,
+						end:   60,
+						text: &bytesRange{
+							start: 49,
+							end:   51,
+						},
+						destination: &bytesRange{
+							start: 53,
+							end:   59,
+						},
+					},
 				},
 			},
-			func(l Link) {
-				if bytes.Equal(l.GetText(), []byte("b")) {
+			func(l Link) (Link, error) {
+				if !bytes.Equal(l.GetText(), []byte("b")) {
 					l.Remove(false)
+					return nil, nil
 				}
+				return l, nil
 			},
-			[]byte("some intro \n`x`some text [b1](c1.com)"),
+			[]byte("some intro [b](c.com)\n`x`some text \n"),
 		},
 		{
 			&document{
-				[]byte("some intro [b](c.com)\n`x`some text [b1](c1.com)"),
+				[]byte("some intro [b](c.com)\n`x`some text [b1](c1.com)\n[b2](c2.com)"),
 				[]Link{
 					&link{
 						start: 11,
@@ -207,14 +224,28 @@ func TestListLinkRewrites(t *testing.T) {
 							end:   46,
 						},
 					},
+					&link{
+						start: 48,
+						end:   60,
+						text: &bytesRange{
+							start: 49,
+							end:   51,
+						},
+						destination: &bytesRange{
+							start: 53,
+							end:   59,
+						},
+					},
 				},
 			},
-			func(l Link) {
-				if bytes.Equal(l.GetText(), []byte("b")) {
+			func(l Link) (Link, error) {
+				if !bytes.Equal(l.GetText(), []byte("b")) {
 					l.Remove(true)
+					return nil, nil
 				}
+				return l, nil
 			},
-			[]byte("some intro b\n`x`some text [b1](c1.com)"),
+			[]byte("some intro [b](c.com)\n`x`some text b1\nb2"),
 		},
 	}
 	for _, tc := range testCases {
@@ -224,7 +255,7 @@ func TestListLinkRewrites(t *testing.T) {
 			}
 			tc.in.ListLinks(tc.listCb)
 			got := tc.in.Bytes()
-			assert.Equal(t, tc.want, got)
+			assert.Equal(t, string(tc.want), string(got))
 		})
 	}
 }

--- a/pkg/markdown/parser/types.go
+++ b/pkg/markdown/parser/types.go
@@ -9,15 +9,18 @@ type Parser interface {
 	Parse(data []byte) Document
 }
 
-// UpdateMarkdownLinkListed is a callback function invoked by the
+// OnLinkListed is a callback function invoked by the
 // ListLinks iterator in Document
-type UpdateMarkdownLinkListed func(link Link)
+type OnLinkListed func(link Link) (Link, error)
 
 // Document is the markdown model parsed from bytes data
 type Document interface {
 	// ListLinks iterates parsed links in this document
-	// and invokes cb on every link
-	ListLinks(cb UpdateMarkdownLinkListed)
+	// and invokes cb on every link. Should the link be
+	// removed, which is signaled by OnLinkListed returning
+	// nil, ListLinks will update the document links list
+	// referenced by each link.
+	ListLinks(cb OnLinkListed) error
 	// Bytes returns the  parsed document content bytes
 	Bytes() []byte
 }

